clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,737 @@
+  => Vec m (BitVector n)
+  -- ^ The Matrix
+  -> BitVector n
+  -- ^ The column vector
+  -> BitVector m
+  -- ^ The resulting vector
+matVecMul mat vec = v2bv $ fmap (\row -> reduceXor $ vec .&. row) mat
+
+-- | Like 'subSNat' but uses a 'Data.Type.Ord.<=' constraint
+subSNatLe
+  :: forall (n :: Nat)
+            (m :: Nat)
+   . m <= n
+  => SNat n
+  -> SNat m
+  -> SNat (n - m)
+subSNatLe a b = leToPlus @m @n $ subSNat a b
+
+-- | Like 'take' but uses a 'Data.Type.Ord.<=' constraint
+takeLe
+  :: forall (n :: Nat)
+            (m :: Nat)
+            (a :: Type)
+   . n <= m
+  => SNat n
+  -- ^ How many elements to take
+  -> Vec m a
+  -- ^ input vector
+  -> Vec n a
+takeLe SNat vs = leToPlus @n @m $ takeI vs
+
+-- | Apply a step to a CRC and input given by @lanePrev@
+--
+-- This function is meant to be used via 'smap' and crashes if called with
+-- out-of-bounds @lanePrev@.
+laneStep
+  :: CrcHardwareParams crcWidth dataWidth nLanes
+  -> SNat lanePrev
+  -> BitVector crcWidth
+  -> Vec nLanes (BitVector dataWidth)
+  -> BitVector crcWidth
+laneStep (CrcHardwareParams SNat nLanes@SNat (CrcParams SNat _ _ _ _ _) laneParams _) lanePrev@SNat crc input
+  = let lane = addSNat lanePrev d1 in case compareSNat lane nLanes of
+    SNatLE -> matVecMul (getFGMatrix laneParams lane) (crc ++# (pack $ takeLe lane input))
+    _      -> clashCompileError "laneStep: Absurd, Report this to the Clash compiler team: https://github.com/clash-lang/clash-compiler/issues"
+
+-- | A parallel multilane CRC engine and validator
+--
+-- @dataWidth@ should /always/ cleanly divide @crcWidth@. If this is not the case
+-- then it is possible for the last message to be misaligned.
+--
+-- For example if you have a @dataWidth@ of @48@ and a @crcWidth@ of @32@. Then a possible
+-- message could be 128 bits, two words + CRC. But this cannot be safely
+-- run through the CRC circuit, because the final fragment contains 32 valid bits
+-- and 16 invalid bits.
+--
+-- This rule is __NOT__ checked at the type level because as long you take
+-- care that misalignment does not occur it is not a problem.
+crcEngine
+  :: forall (dom :: Domain)
+            (crc :: Type)
+            (dataWidth :: Nat)
+            (nLanes :: Nat)
+   . HiddenClockResetEnable dom
+  => HardwareCrc crc dataWidth nLanes
+  => crc
+  -- ^ The CRC
+  -> Signal dom (Maybe (Bool, Index nLanes, Vec nLanes (BitVector dataWidth)))
+  -- ^ The input data.
+  --   The @Bool@ must be asserted on the first fragment to start a new CRC
+  --   computation
+  --   @Index nLanes@ indicates how many @dataWidth@ words are
+  --   valid in the vector minus 1.
+  --
+  --     - Ex. 1. @Just (False, 0, 0xDE :> 0xAD :> 0xBE :> 0xEF :> Nil)@
+  --       means this is not the start of a new message and only the first word @0xDE@ is valid
+  --     - Ex. 2. @Just (True, 3, 0xDE :> 0xAD :> 0xBE :> 0xEF :> Nil@
+  --       means this is the start of a new message and all words in the vector are valid
+  --
```

Comment:
```suggestion
  -- ^ The input data
  --
  --   The @Bool@ must be asserted on the first fragment to start a new CRC
  --   computation.
  --
  --   @Index nLanes@ indicates how many @dataWidth@ words are valid in the
  --   vector minus 1.
  --
  --     - Ex. 1. @Just (False, 0, 0xDE :> 0xAD :> 0xBE :> 0xEF :> Nil)@
  --       means this is not the start of a new message and only the first word
  --       @0xDE@ is valid.
  --     - Ex. 2. @Just (True, 3, 0xDE :> 0xAD :> 0xBE :> 0xEF :> Nil@ means
  --       this is the start of a new message and all words in the vector are
  --       valid.
```

If you want multiple paragraphs, you'll need to leave a blank line in between.

I made some more tweaks: periods, trailing blank line, line length.